renderdoc/driver/d3d12/d3d12_counters.cpp

```diff
@@ -228,106 +241,257 @@ vector<CounterResult> D3D12Replay::FetchCounters(const vector<uint32_t> &counter
+    return ret;
+  }
 
-  D3D12GPUTimerCallback cb(m_pDevice, this, queryHeap);
+  D3D12_QUERY_HEAP_DESC pipestatsQueryDesc;
+  pipestatsQueryDesc.Count = maxEID;
+  pipestatsQueryDesc.NodeMask = 1;
+  pipestatsQueryDesc.Type = D3D12_QUERY_HEAP_TYPE_PIPELINE_STATISTICS;
+  ID3D12QueryHeap *pipestatsQueryHeap = NULL;
+  hr = m_pDevice->CreateQueryHeap(&pipestatsQueryDesc, __uuidof(pipestatsQueryHeap),
+                                  (void **)&pipestatsQueryHeap);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create pipeline statistics query heap %08x", hr);
+    return ret;
+  }
+
+  D3D12_QUERY_HEAP_DESC occlusionQueryDesc;
+  occlusionQueryDesc.Count = maxEID;
+  occlusionQueryDesc.NodeMask = 1;
+  occlusionQueryDesc.Type = D3D12_QUERY_HEAP_TYPE_OCCLUSION;
+  ID3D12QueryHeap *occlusionQueryHeap = NULL;
+  hr = m_pDevice->CreateQueryHeap(&occlusionQueryDesc, __uuidof(occlusionQueryHeap),
+                                  (void **)&occlusionQueryHeap);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create occlusion query heap %08x", hr);
+    return ret;
+  }
+
+  // Only supported with developer mode drivers!!!
+  // m_pDevice->SetStablePowerState(TRUE);
+
+  D3D12GPUTimerCallback cb(m_pDevice, this, timerQueryHeap, pipestatsQueryHeap, occlusionQueryHeap);
 
   // replay the events to perform all the queries
   m_pDevice->ReplayLog(0, maxEID, eReplay_Full);
 
-  m_pDevice->SetStablePowerState(FALSE);
+  // Only supported with developer mode drivers!!!
+  // m_pDevice->SetStablePowerState(FALSE);
 
   ID3D12GraphicsCommandList *list = m_pDevice->GetNewList();
 
-  list->ResolveQueryData(queryHeap, D3D12_QUERY_TYPE_TIMESTAMP, 0, maxEID * 2, readbackBuf, 0);
+  list->ResolveQueryData(timerQueryHeap, D3D12_QUERY_TYPE_TIMESTAMP, 0, maxEID * 2,
+                         timerReadbackBuf, 0);
+  list->ResolveQueryData(pipestatsQueryHeap, D3D12_QUERY_TYPE_PIPELINE_STATISTICS, 0, maxEID,
+                         pipestatsReadbackBuf, 0);
+  list->ResolveQueryData(occlusionQueryHeap, D3D12_QUERY_TYPE_OCCLUSION, 0, maxEID,
+                         occlusionReadbackBuf, 0);
 
   list->Close();
 
   m_pDevice->ExecuteLists();
   m_pDevice->FlushLists();
 
-  D3D12_RANGE range = {0, (SIZE_T)bufDesc.Width};
-  void *data;
-  hr = readbackBuf->Map(0, &range, &data);
+  D3D12_RANGE range = {0, (SIZE_T)timerBufDesc.Width};
+  void *timerData;
+  hr = timerReadbackBuf->Map(0, &range, &timerData);
   if(FAILED(hr))
   {
-    RDCERR("Failed to create query heap %08x", hr);
-    SAFE_RELEASE(queryHeap);
-    SAFE_RELEASE(readbackBuf);
+    RDCERR("Failed to read timer query heap data %08x", hr);
+    SAFE_RELEASE(timerQueryHeap);
+    SAFE_RELEASE(timerReadbackBuf);
+    SAFE_RELEASE(pipestatsQueryHeap);
+    SAFE_RELEASE(pipestatsReadbackBuf);
+    SAFE_RELEASE(occlusionQueryHeap);
+    SAFE_RELEASE(occlusionReadbackBuf);
     return ret;
   }
 
-  uint64_t *timestamps = (uint64_t *)data;
+  range = {0, (SIZE_T)pipestatsBufDesc.Width};
```

Comment:
Yep, I use VS2010 so that's the minimum compiler version for the code to compile.